### ISSUE TITLE
Concrete payment details in PaymentComponentDelegate. Custom PaymentMethodDetails Enconding.

### ISSUE THREAD

**Issue by demalex** [Adyen/adyen-ios#104]
Hi @joostvandijk,

Because of PCI Compliance we are obliged to use `CardComponent` provided by Adyen. In order to keep Adyen framework decoupled from application and domain specific code we need to map `CardDetails` DTO to domain specific DTO. Currently this is not possible, because `PaymentComponentDelegate` returns an abstraction 

`PaymentComponentData.PaymentMethodDetails` and those fields are not accessible.
As an option it is possible to use plain `Encodable`, but it means that we can't apply specific encoding which could be required by our merchant backend.

Is there any options to introduce a concrete payment object in `PaymentComponentDelegate `?
I understand that this breaks a `PaymentComponentDelegate` as an abstraction, but maybe you have any other ideas in mind?

Anyway, any input from your side will be appreciated!

Thank you!

**Comment by joostvandijk** [comment]
Hi @demalex,

I'm not 100% if I understand your issue correctly. `PaymentComponentDelegate` indeed returns an abstract `PaymentMethodDetails` interface, but you can cast this to `CardDetails` and read out the information that you need. If that doesn't solve your issue, do you have some code suggestions that would help you fix this issue?

With kind regards,

Joost van Dijk

**Comment by demalex** [comment]
Hi @joostvandijk,

Please, sorry for the delay with reply!

Thank you very much for your suggestion, that is what we are doing right now. 

From my perspective there is no point to have interface, if then it needs to be casted to concrete type. That is why I asked for your input, might be that I was missing something.

Another option could be to use type as `Encodable` or `[String: Any]` . But non of those types could be used by `Encoder` as is. 

Of course, we can work around it creating kinda `AnyEncodable` or extensions for `KeyedEncodingContainer` as you did in `AdyenUIHost`, but it would be great to have third party library ready to use from the very beginning.

Please, let me know if you have any additional suggestions how we can make it more convenient to use.

Thanks!

**Comment by joostvandijk** [comment]
Hi @demalex,

Thanks for the detailed reply. This is indeed still an issue we have yet to figure out. We want to reuse the same delegate for every component, because it eases integration when using multiple components. This is also why `PaymentMethodDetails` is a protocol. However, this does cause issues with `Encoder`.

We are looking into using some of Swift 5's new features to improve this setup. If you have any ideas/recommendations, please let us know!

Best,

Joost

**Comment by mohammedDehairy** [comment]
Hi @demalex,

 This issue should be solved with 3.4.0+ versions, as now the PaymentMethodDetails can be encoded without access to the concrete type through calling:

```
try container.encode(data.paymentMethod.encodable, forKey: .details)
```
please look at the example in `PaymentsRequest.swift`.

Does this solve your problem?

Mohamed

**Comment by demalex** [comment]
Hi @mohammedDehairy,

Thank you for your message reply and I apologize for the delay with my reply.

As far as I remember there was no problem to serialize it. That moment we could use `dictionaryRepresentation` and get `Data` of  it using `JSONSerialization`. 

What we wanted is as follows:

1. Possibility to define custom serialization. e.g. we want to use JSON keys which are different from encodable object property names. One reason for that is that we want to keep our own API contract with backend. By using default serialization we are introducing implicit dependency, because now contract is enforced by SDK. If output from is  `Data` or `[String: Any]` it becomes fragile, not typed and enforce by compiler. So, potential problem could be identified on application testing phases later.

2. Have an access to concrete object instead of `PaymentMethodDetails` in order to be able to map it to application specific DTO (also, we do not need `storedPaymentMethodIdentifier` field). This way we do not introduce coupling by passing third party entity all over the code base. And if something is changed on library side, we can fix it in one mapping place.

Moreover `CardDetails` is public, non mutable struct with encrypted fields. 

So, from our perspective there is no much sense to keep it abstract under `PaymentMethodDetails` protocol. 

On our side we solved this by casting `PaymentMethodDetails` to `CardDetails` with `assertionFailure` if casting fails. 

Thank you!
